packages/bsky/src/notifications.ts

```diff
@@ -0,0 +1,334 @@
+        // If batch size is 20, add task to background queue
+        if (this.notificationBatch.length >= this.notificationBatchSize) {
+          const batch = [...this.notificationBatch]
+          this.notificationBatch = []
+          this.backgroundQueue.add(async () => {
+            try {
+              await this.sendPushNotifications(batch)
+            } catch (err) {
+              logger.error({ err, batch }, 'notification push batch failed')
+            }
+          })
+        }
+      }
+    }
+  }
+
+  /**  1. Get the user's token (APNS or FCM for iOS and Android respectively) from the database
+    User token will be in the format:
+        did || token || platform (1 = iOS, 2 = Android, 3 = Web)
+    2. Send notification to `gorush` server with token
+    Notification will be in the format:
+    "notifications": [
+      {
+        "tokens": string[],
+        "platform": 1 | 2,
+        "message": string,
+        "title": string,
+        "priority": "normal" | "high",
+        "image": string, (Android only)
+        "expiration": number, (iOS only)
+        "badge": number, (iOS only)
+      }
+    ]
+    3. `gorush` will send notification to APNS or FCM
+    4.  store response from `gorush` which contains the ID of the notification
+    5. If notification needs to be updated or deleted, find the ID of the notification from the database and send a new notification to `gorush` with the ID (repeat step 2)
+  */
+  private async sendPushNotifications(notifications: PushNotification[]) {
+    // if pushEndpoint is not defined, we are not running in the indexer service, so we can't send push notifications
+    if (!this.pushEndpoint) {
+      throw new Error('Push endpoint not defined')
+    }
+    // if no notifications, skip and return early
+    if (notifications.length === 0) {
+      return
+    }
+    await axios.post(
+      this.pushEndpoint,
+      { notifications },
+      {
+        headers: {
+          'Content-Type': 'application/json',
+          accept: 'application/json',
+        },
+      },
+    )
+  }
+
+  async registerDeviceForPushNotifications(
+    did: string,
+    token: string,
+    platform: Platform,
+    appId: string,
+  ) {
+    // if token doesn't exist, insert it, on conflict do nothing
+    await this.db.db
+      .insertInto('notification_push_token')
+      .values({ did, token, platform, appId })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+  }
+
+  async getNotificationDisplayAttributes(
+    notifs: InsertableNotif[],
+  ): Promise<NotifDisplay[]> {
+    const { ref } = this.db.db.dynamic
+    const authorDids = notifs.map((n) => n.author)
+    const subjectUris = notifs.flatMap((n) => n.reasonSubject ?? [])
+    const recordUris = notifs.map((n) => n.recordUri)
+    const allUris = [...subjectUris, ...recordUris]
+
+    // gather potential display data for notifications in batch
+    const [authors, posts] = await Promise.all([
```

Comment:
i still wonder if we can avoid most of these db hits by passing the record into the notification service?

i will see - these bulk gets give a decent justification for processing notifs in batches

Reply:
I believe that does help with the post record lookup for mentions and replies, which do show the text from the post that was just indexed.  The author display name lookups for all notifs and post record lookups for likes, quotes, reposts will still need to happen at some point.  The changes I made here tried to batch work together into two queries rather than a few queries per notification.  Batching more notifications together as you're suggesting will probably be what gives the biggest win here.